src/flowsynth.py

```diff
@@ -466,25 +465,56 @@ def _valid_mac(self, mac):
     #This function expects all inputs to be enclosed within double quotes
     def parse_content(self, content):
         """ parse and render a content keyword """
+
+        # this regex is somewhat gnarly but leaving for now ... (why not
+        # just strip off double quotes on ends)?
         pcre_text = r'"([^\\"]*(?:\\.[^\\"]*)*)"'
 
+        result = bytearray()
 
         #first, check for text
         mo_text = re.match(pcre_text, content)
         if (mo_text != None):
-            logging.debug("Content: %s", mo_text.group(1))
-
             content_text = mo_text.group(1)
-            replacements = re.findall(r"\\x[a-fA-F0-9]{2}", content_text)
-            for replacement in replacements:
-                content_text = content_text.replace(replacement, chr(int(replacement[2:], 16)))
+            logging.debug("Content: %s (length %d)" % (content_text, len(content_text)))
+            start = 0
+            previous_end = 0
+            # Flowsynth supports encoding arbitrary bytes with the "\xHH" notation where "HH" is
+            # the hexidecimal representation of the byte. That is what is handled here, while
+            # maintaining the rest of the content data as UTF-8.
+            for hex_replacement in re.finditer(r"\\x[a-fA-F0-9]{2}", content_text):
+                # try/catch blocks to deal with different data representation from shlex (depends on Python version)
+                start = hex_replacement.start(0)
+                end = hex_replacement.end(0)
+                ascii_hex = content_text[start+2:start+4]
+                previous_substring = content_text[previous_end:start]
+                if len(previous_substring) > 0:
+                    # extend result with previous substring; encode as UTF-8
+                    try:
+                        result.extend(previous_substring.encode('utf-8'))
+                    except UnicodeDecodeError:
+                        result.extend(previous_substring)
+                # append ASCII hex byte to result
+                result.extend(bytearray.fromhex(ascii_hex))
+                previous_end = end
+            if previous_end == 0:
+                # no hex encoding found, just encode the whole thing
+                try:
+                    result.extend(content_text.encode('utf-8'))
+                except UnicodeDecodeError:
+                    result.extend(content_text)
+            elif previous_end < len(content_text):
+                # add the last substring
+                try:
+                    result.extend(content_text[previous_end:len(content_text)].encode('utf-8'))
+                except UnicodeDecodeError:
+                    result.extend(content_text[previous_end:len(content_text)])
```

Comment:
It strikes me that this is the sort of code for which unit tests have a big payoff.  It would be nice if there were a stable of example inputs and outputs that demonstrated what you want this code to do so you can be confident that it's handling all the cases you have in mind.

Reply:
And since it'll need to be shown to work in py2 and py3 the use of tox to exercise both is something you should consider.